service/family_tree_test.go

```diff
@@ -0,0 +1,473 @@
+package familytree
+
+import (
+	"errors"
+	"testing"
+)
+
+func TestAddNode(t *testing.T) {
+	ft := NewFamilyTree()
+	tests := []struct {
+		scenario string
+		id       int
+		name     string
+		metadata map[string]string
+		err      error
+	}{
+		{
+			scenario: "node gets added successfully",
+			id:       1,
+			name:     "A",
+			metadata: nil,
+			err:      nil,
+		},
+		{
+			scenario: "node do not get added as it is repeated",
+			id:       1,
+			name:     "A",
+			metadata: nil,
+			err:      errors.New("node already exists"),
+		},
+	}
+
+	for _, tc := range tests {
+		err := ft.AddNode(tc.id, tc.name, tc.metadata)
+		if err != nil && tc.err == nil {
+			t.Errorf("for scenario: %s, expected %v, got %v", tc.scenario, tc.err, err)
+		}
+		if err == nil && tc.err != nil {
+			t.Errorf("for scenario: %s, expected %v, got %v", tc.scenario, tc.err, err)
+		}
+	}
+}
+
+func TestAddEdge(t *testing.T) {
+	ft := NewFamilyTree()
+	ft.AddNode(1, "A", nil)
+	ft.AddNode(2, "B", nil)
+	ft.AddNode(3, "C", nil)
+	ft.AddNode(4, "D", nil)
+	ft.AddNode(5, "E", nil)
+	ft.AddEdge(1, 2)
+	ft.AddEdge(5, 2)
+	ft.AddEdge(2, 3)
+	ft.AddEdge(4, 3)
+
+	tests := []struct {
+		scenario string
+		childID  int
+		parentID int
+		err      error
+	}{
+		{
+			scenario: "a valid edge",
+			childID:  2,
+			parentID: 1,
+			err:      nil,
+		},
+		{
+			scenario: "a valid edge adding again",
+			childID:  2,
+			parentID: 1,
+			err:      nil,
+		},
+		{
+			scenario: "invalid edge parent and child same",
+			childID:  2,
+			parentID: 2,
+			err:      errors.New("parent and child have same id"),
+		},
+		{
+			scenario: "cyclic case: child trying to become a parent",
+			childID:  1,
+			parentID: 2,
+			err:      errors.New("cyclic case"),
+		},
+		{
+			scenario: "parent id not exist",
+			childID:  1,
+			parentID: 7,
+			err:      errors.New("parent id not exist"),
+		},
+		{
+			scenario: "child id not exist",
+			childID:  9,
+			parentID: 7,
+			err:      errors.New("parent and child id not exist"),
+		},
+	}
+	for _, tc := range tests {
+		err := ft.AddEdge(tc.parentID, tc.childID)
+		if tc.err != nil && err == nil {
```

Comment:
ft state changes with every tests
have it defined as part of single test as prerequisite for every test